### PR TITLE
Add edge-case coverage for the dotnet test wire protocol serializers

### PR DESCRIPTION
**Pull request by Evangelink**
## What

Pure **test additions** complementing `ProtocolTests` with edge cases for the `dotnet test` wire-protocol NamedPipe serializers. Independent of the terminal-reporter PRs (base `main`).

Covers:
- `TestResultMessages`: both lists empty; multiple successful + failed in order; null optional fields; failed result with no exceptions; **multiple exceptions** per failed test.
- **Special characters** (unicode, CRLF, tab, NUL, quotes) round-trip exactly.
- Empty lists for `DiscoveredTestMessages` / `FileArtifactMessages` / `CommandLineOptionMessages`.
- `FileArtifactMessages` with multiple artifacts + mixed null fields.
- `TestInProgressMessage` all-null; `TestSessionEvent` end event.

## Verification
`ProtocolTests` + `ProtocolEdgeCaseTests` green: **26 total, 0 failed**.

### CODE REVIEW

File: test/UnitTests/Microsoft.Testing.Platform.UnitTests/IPC/ProtocolEdgeCaseTests.cs [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+    private static TMessage RoundTrip<TMessage>(object serializer, TMessage message)
+    {
+        var stream = new MemoryStream();
+        Serialize(serializer, message, stream);
+        stream.Seek(0, SeekOrigin.Begin);
+        return (TMessage)Deserialize(serializer, stream);
+    }
+
+    private static void Serialize<TMessage>(object serializer, TMessage message, Stream stream)
```
Comment (Evangelink) [inline test/UnitTests/Microsoft.Testing.Platform.UnitTests/IPC/ProtocolEdgeCaseTests.cs]: **[MODERATE] Code Structure & Simplification — Verbatim duplication of helper infrastructure**

`Serialize<TMessage>` (lines 180–184) and `Deserialize` (lines 186–190) are character-for-character copies of the identical private helpers that already exist in `ProtocolTests` (lines 433–443 of that file). Both files live in the same namespace (`Microsoft.Testing.Platform.UnitTests`), so sharing the helpers is trivial.

**Concrete maintenance risk:** both helpers use a reflection-based `Single(...)` that throws `InvalidOperationException` if the serializer's method signatures ever change (e.g., an overload added, a parameter renamed). When that happens the author must find and update both files. Because the helpers are identical, a divergent edit is practically inevitable over time.

**Suggested fix:** Extract to a `file`-scoped or `internal static` helper class in the same directory:

```csharp
// ProtocolSerializerTestHelpers.cs (new file, same namespace)
internal static class ProtocolSerializerTestHelpers
{
    internal static TMessage RoundTrip<TMessage>(object serializer, TMessage message)
    {
        using var stream = new MemoryStream();
        Serialize(serializer, message, stream);
        stream.Seek(0, SeekOrigin.Begin);
        return (TMessage)Deserialize(serializer, stream);
    }

    internal static void Serialize<TMessage>(object serializer, TMessage message, Stream stream)
        => serializer.GetType()
            .GetMethods(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic)
            .Single(m => m.Name == nameof(Serialize)
                && m.GetParameters() is [{ ParameterType: var mt }, { ParameterType: var st }]
                && mt == typeof(TMessage) && st == typeof(Stream))
            .Invoke(serializer, [message!, stream]);

    internal static object Deserialize(object serializer, Stream stream)
        => serializer.GetType()
            .GetMethods(BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic)
            .Single(m => m.Name == nameof(Deserialize)
                && m.GetParameters() is [{ ParameterType: var pt }]
                && pt == typeof(Stream))
            .Invoke(serializer, [stream])!;
}
```

`ProtocolTests` and `ProtocolEdgeCaseTests` both delegate to it. The `using var stream` also fixes the minor `MemoryStream`-not-disposed pattern while we're here (though `MemoryStream` has no unmanaged resources, the `using` is idiomatic and satisfies the Disposable Management dimension).

File: test/UnitTests/Microsoft.Testing.Platform.UnitTests/IPC/ProtocolEdgeCaseTests.cs [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    public void TestResultMessages_WhenBothListsEmpty_RoundTrips()
+    {
+        var message = new TestResultMessages("exec", "inst", [], []);
+
+        TestResultMessages actual = RoundTrip(new TestResultMessagesSerializer(), message);
+
+        Assert.AreEqual("exec", actual.ExecutionId);
+        Assert.AreEqual("inst", actual.InstanceId);
+        Assert.IsEmpty(actual.SuccessfulTestMessages);
+        Assert.IsEmpty(actual.FailedTestMessages);
```
Comment (Evangelink) [inline test/UnitTests/Microsoft.Testing.Platform.UnitTests/IPC/ProtocolEdgeCaseTests.cs:29]: **[NIT] Assertion Quality — `InstanceId` set but never asserted**

`InstanceId` is passed as `"inst"` in the constructor (line 20) but none of the assertions on lines 24–27 verify that it round-trips correctly. The analogous tests in `ProtocolTests` always assert both `ExecutionId` and `InstanceId` (e.g., `DiscoveredTestMessagesSerializeDeserialize` lines 55–56).

The same gap exists in `DiscoveredTestMessages_WhenEmptyList_RoundTrips` (line 101) and `FileArtifactMessages_WhenEmptyList_RoundTrips` (line 111) — both set `"inst"` but only assert `ExecutionId`.

A bug that zero-initialised `InstanceId` on deserialization would not be caught by any of these three tests.

```csharp
Assert.AreEqual("inst", actual.InstanceId);
```

**Review by Evangelink — COMMENTED** [review]
> [!NOTE]
> 🤖 **Automated review by GitHub Copilot.** Posted via a maintainer's GitHub token, so it appears under their account — the account owner did **not** write or approve this content personally. Generated by the [Expert Code Review workflow](https://github.com/microsoft/testfx/actions/runs/27843212491). To request a follow-up action, reply by tagging `@copilot` directly.

## Expert Review — `ProtocolEdgeCaseTests.cs`

The test coverage itself is solid. The scenarios selected (empty collections, multiple items, multi-exception, null optional fields, special characters, all-null fields, `TestSessionEnd`) are meaningful complements to `ProtocolTests`. Assertion style is correct: MSTest `Assert`/`CollectionAssert` throughout, matching the `BannedSymbols.txt` policy that bans `AwesomeAssertions` in this project. No threading, security, or flakiness concerns.

One concrete MODERATE finding blocks the review: the test helpers are duplicated verbatim from `ProtocolTests`. See the inline comment on line 180.

---

### 22-Dimension Verdict Table

| # | Dimension | Status | Notes |
|---|-----------|--------|-------|
| 1 | Algorithmic Correctness | ✅ LGTM | Round-trip helpers are correct; no off-by-one or logic inversion |
| 2 | Threading & Concurrency | N/A | Test file; no shared mutable state |
| 3 | Security & IPC Contract Safety | N/A | No production IPC code changed |
| 4 | Public API & Binary Compatibility | N/A | No public API surface changes |
| 5 | Performance & Allocations | N/A | Per-test reflection is acceptable in a test |
| 6 | Cross-TFM Compatibility | ✅ LGTM | No TFM-gated APIs used |
| 7 | Resource & IDisposable Management | ⚠️ NIT | `MemoryStream` in `RoundTrip` (line 174) created without `using`; matches the existing pattern in `ProtocolTests` but the new extracted helper is a good place to fix it (folded into the D15 recommendation) |
| 8 | Defensive Coding at Boundaries | N/A | — |
| 9 | Localization & Resources | N/A | — |
| 10 | Test Isolation | ✅ LGTM | No static mutable state, no fixed file paths, no environment mutation |
| 11 | Assertion Quality | ⚠️ NIT | `InstanceId` is set in three tests but never asserted on the round-tripped value; see inline comment on line 27 |
| 12 | Flakiness Patterns | ✅ LGTM | No `Thread.Sleep`, no timing, no hard-coded ports, no duration regexes |
| 13 | Test Completeness & Coverage | ✅ LGTM | Good scenario variety; `\u0000` and `\u2728` cover the binary encoding edge cases; the serializer uses `Encoding.UTF8.GetBytes(string)` (not char-by-char), so surrogate pairs are handled correctly by the production code — no gap there |
| 14 | Data-Driven Test Coverage | N/A | No `[DataRow]`/`[DynamicData]` tests |
| 15 | Code Structure & Simplification | ⛔ **MODERATE** | `Serialize<TMessage>` and `Deserialize` (lines 180–190) are verbatim copies of the identically-named helpers in `ProtocolTests`; see inline comment on line 180 |
| 16 | Naming & Conventions | ⚠️ NIT | New tests use `Method_WhenCondition_Result` (BDD) naming; `ProtocolTests` uses `MethodNameScenario` (flat camelCase). Both styles are acceptable — the BDD form is actually more descriptive — but the inconsistency within the IPC test folder is worth noting. Not a blocker. |
| 17 | Documentation Accuracy | ✅ LGTM | Class-level `<summary>` accurately describes the scope and the cross-reference to `ProtocolTests` |
| 18 | Analyzer & Code Fix Quality | N/A | No analyzer changes |
| 19 | IPC Wire Compatibility | ✅ LGTM | Tests validate the wire format; no serializer IDs or field IDs are changed |
| 20 | Build Infrastructure & Dependencies | N/A | — |
| 21 | Scope & PR Discipline | ✅ LGTM | Single-concern PR (new tests only), clean commit |
| 22 | PowerShell Scripting Hygiene | N/A | No `.ps1` changes |

---

### Summary

**1 MODERATE finding** (D15 — duplicate helpers, inline comment on line 180) and **2 NITs** (D7/D11 — `MemoryStream` disposal can be fixed as part of the helper extraction, and three missing `InstanceId` assertions). The MODERATE finding has a straightforward fix and the suggested extraction also cleans up the NIT. All other dimensions are clean.

File: test/UnitTests/Microsoft.Testing.Platform.UnitTests/IPC/ProtocolEdgeCaseTests.cs (hunk 1 above)
Comment (Evangelink) [reply test/UnitTests/Microsoft.Testing.Platform.UnitTests/IPC/ProtocolEdgeCaseTests.cs]: Done in 8d96ce8a4 — extracted the reflection-based `Serialize`/`Deserialize`/`RoundTrip` into a shared `ProtocolSerializerTestHelper` (same namespace). Both `ProtocolTests` and `ProtocolEdgeCaseTests` now delegate to it via `using static`, and the duplicated private helpers were removed from `ProtocolTests`.

File: test/UnitTests/Microsoft.Testing.Platform.UnitTests/IPC/ProtocolEdgeCaseTests.cs (hunk 2 above)
Comment (Evangelink) [reply test/UnitTests/Microsoft.Testing.Platform.UnitTests/IPC/ProtocolEdgeCaseTests.cs:29]: Done in 8d96ce8a4 — added `Assert.AreEqual(\inst\, actual.InstanceId)` to the three round-trip tests that previously only asserted `ExecutionId` (`TestResultMessages_WhenBothListsEmpty_RoundTrips`, `DiscoveredTestMessages_WhenEmptyList_RoundTrips`, `FileArtifactMessages_WhenEmptyList_RoundTrips`).
